### PR TITLE
Terrain height probe

### PR DESCRIPTION
**Pull request by flyingk**
This adds the capability to obtain the terrain height above MSL at a given location using a XPLM terrain probe. A new message pair (GETT query, terrain response) was added to specify the probe location and to return the data.

The probe is persistent once created for speed. Currently only one probe is supported.

Using this mechanism allows to land an aircraft onto the x-plane world.

Definitions for the message formats will be added shortly.

### CODE REVIEW

**Comment by flyingk** [comment]
message formats:
[xplane_terrain_msg.xlsx](https://github.com/nasa/XPlaneConnect/files/3814245/xplane_terrain_msg.xlsx)

File: xpcPlugin/MessageHandlers.cpp [hunk 1]
```diff
@@ -66,6 +69,7 @@ namespace XPC
 			handlers.insert(std::make_pair("VIEW", MessageHandlers::HandleView));
 			handlers.insert(std::make_pair("GETC", MessageHandlers::HandleGetC));
 			handlers.insert(std::make_pair("GETP", MessageHandlers::HandleGetP));
+			handlers.insert(std::make_pair("GETT", MessageHandlers::HandleGetT)); // terrain properties
```
Comment (jason-watkins) [inline xpcPlugin/MessageHandlers.cpp]: Remove comment

File: xpcPlugin/MessageHandlers.cpp [hunk 2]
```diff
@@ -626,6 +630,89 @@ namespace XPC
...
+		std::size_t size = msg.GetSize();
+		if (size != 30)
+		{
+			Log::FormatLine(LOG_ERROR, "GTER", "Unexpected message length: %u", size);
+			return;
+		}
+		unsigned char aircraft = buffer[5];
+		Log::FormatLine(LOG_TRACE, "GTER", "Getting terrain information for aircraft %u", aircraft);
+		
+		double loc[3], X, Y, Z;
```
Comment (jason-watkins) [inline xpcPlugin/MessageHandlers.cpp]: This is really personal preference, but I don't care for this style of declaration. Please declare one thing per line.

File: xpcPlugin/MessageHandlers.cpp [hunk 3]
```diff
@@ -626,6 +630,89 @@ namespace XPC
...
 		}
 	}
+	
+	void MessageHandlers::HandleGetT(const Message& msg)
+	{
+		const unsigned char* buffer = msg.GetBuffer();
+		std::size_t size = msg.GetSize();
+		if (size != 30)
+		{
+			Log::FormatLine(LOG_ERROR, "GTER", "Unexpected message length: %u", size);
```
Comment (jason-watkins) [inline xpcPlugin/MessageHandlers.cpp]: "GTER" should be "MSGH". The main point of those tags is to quickly indicate which source file generated a particular line in the log.

File: xpcPlugin/MessageHandlers.cpp [hunk 4]
```diff
@@ -626,6 +630,89 @@ namespace XPC
...
+			return;
+		}
+		unsigned char aircraft = buffer[5];
+		Log::FormatLine(LOG_TRACE, "GTER", "Getting terrain information for aircraft %u", aircraft);
```
Comment (jason-watkins) [inline xpcPlugin/MessageHandlers.cpp]: Change log tag

File: xpcPlugin/MessageHandlers.cpp [hunk 5]
```diff
@@ -626,6 +630,89 @@ namespace XPC
...
+			XPLMWorldToLocal(loc[0], loc[1], loc[2], &X, &Y, &Z);
+		}
+		
+		// Init terrain probe (if required) and probe data struct
+		XPLMProbeInfo_t probe_data;
+		probe_data.structSize = sizeof(XPLMProbeInfo_t);
+		
+		if(Terrain_probe == nullptr)
+		{
+			Log::FormatLine(LOG_TRACE, "GTER", "Create terrain probe for aircraft %u", aircraft);
```
Comment (jason-watkins) [inline xpcPlugin/MessageHandlers.cpp]: Change log tag

File: xpcPlugin/MessageHandlers.cpp [hunk 6]
```diff
@@ -626,6 +630,89 @@ namespace XPC
...
+			Terrain_probe = XPLMCreateProbe(0);
+		}
+		
+		// query probe
+		int rc = XPLMProbeTerrainXYZ(Terrain_probe, X, Y, Z, &probe_data);
+		
+		// transform probe location to world coordinates
+		double lat, lon, alt;
```
Comment (jason-watkins) [inline xpcPlugin/MessageHandlers.cpp]: Split declarations.

File: xpcPlugin/MessageHandlers.cpp [hunk 7]
```diff
@@ -626,6 +630,89 @@ namespace XPC
...
+		
+		if(rc == 0)
+		{
+			XPLMLocalToWorld(probe_data.locationX, probe_data.locationY, probe_data.locationZ, &lat, &lon, &alt);
+			
+			Log::FormatLine(LOG_TRACE, "GTER", "Probe LLA %lf %lf %lf", lat, lon, alt);
```
Comment (jason-watkins) [inline xpcPlugin/MessageHandlers.cpp]: Change log tag

**Comment by jason-watkins** [comment]
Also, I don't suppose you have any idea why this PR seems to break the macOS build on Azure?

File: xpcPlugin/MessageHandlers.cpp (hunk 3 above)
Comment (jason-watkins) [reply xpcPlugin/MessageHandlers.cpp]: I should note that I myself have hardly been consistent on this point, so it's hardly your fault. That said, I'd like to move towards one tag per file for logging.

Comment (flyingk) [reply xpcPlugin/MessageHandlers.cpp]: All functions in MessageHandler have their name in that tag. If using MSGH everywhere, one would loose any pointer which functionality is affected? Can change it, but to me the current way makes more sense.

Comment (jason-watkins) [reply xpcPlugin/MessageHandlers.cpp]: Alright, forget changing it to `MSGH`. Should it at least match the message name (i.e. `GETT` instead of `GTER`)?

**Comment by flyingk** [comment]
The error message is about not finding Mac OS SDK 10.14. Is it running on an older system/Xcode version? 10.14 is specified in that yml file. Maybe try to set it to earlier version? Builds/works fine on 10.14 for me.

**Comment by jason-watkins** [comment]
Well, using the 10.15 SDK seems to work, but everything I can find in the Azure docs suggests that 10.14 should continue to work as well. Anyway, I think it's fixed now and I'm not aware of any downside to building the plugin against the latest SDK version.

File: xpcPlugin/MessageHandlers.cpp (hunk 3 above)
Comment (flyingk) [reply xpcPlugin/MessageHandlers.cpp]: I'll change it to GETT.

**Comment by flyingk** [comment]
Would you be able to add the message format (as provided in comment above) to the wiki? Can't send PR for that...

**Comment by jason-watkins** [comment]
Wiki updated.

**Pull request merged**